File: go/code_smells/dispensables/duplicated_code.go
package dispensables

import "fmt"

// EmailService demonstrates duplicated code smell
type EmailService struct{}

// SendWelcomeEmail sends a welcome email
func (es *EmailService) SendWelcomeEmail(email, name string) error {
	// Duplicated validation
	if email == "" {
		return fmt.Errorf("email is required")
	}
	if !containsAt(email) {
		return fmt.Errorf("invalid email format")
	}

	// Duplicated formatting
	subject := "Welcome!"
	body := fmt.Sprintf("Hello %s, welcome to our service!", name)

	// Duplicated sending logic
	fmt.Printf("Sending email to: %s\n", email)
	fmt.Printf("Subject: %s\n", subject)
	fmt.Printf("Body: %s\n", body)
	fmt.Println("Email sent successfully")

	return nil
}

// SendPasswordResetEmail sends a password reset email
func (es *EmailService) SendPasswordResetEmail(email, resetToken string) error {
	// Duplicated validation
	if email == "" {
		return fmt.Errorf("email is required")
	}
	if !containsAt(email) {
		return fmt.Errorf("invalid email format")
	}

	// Duplicated formatting
	subject := "Password Reset"
	body := fmt.Sprintf("Use this token to reset your password: %s", resetToken)

	// Duplicated sending logic
	fmt.Printf("Sending email to: %s\n", email)
	fmt.Printf("Subject: %s\n", subject)
	fmt.Printf("Body: %s\n", body)
	fmt.Println("Email sent successfully")

	return nil
}

// SendOrderConfirmationEmail sends an order confirmation email
func (es *EmailService) SendOrderConfirmationEmail(email, orderID string, amount float64) error {
	// Duplicated validation
	if email == "" {
		return fmt.Errorf("email is required")
	}
	if !containsAt(email) {
		return fmt.Errorf("invalid email format")
	}

	// Duplicated formatting
	subject := "Order Confirmation"
	body := fmt.Sprintf("Your order %s for $%.2f has been confirmed", orderID, amount)

	// Duplicated sending logic
	fmt.Printf("Sending email to: %s\n", email)
	fmt.Printf("Subject: %s\n", subject)
	fmt.Printf("Body: %s\n", body)
	fmt.Println("Email sent successfully")

	return nil
}

func containsAt(email string) bool {
	for _, c := range email {
		if c == '@' {
			return true
		}
	}
	return false
}

// Exercise: Refactor this code to eliminate duplication
// 1. Extract the common validation logic
// 2. Extract the common email sending logic
// 3. Create a generic sendEmail method that the others can use

File: java/src/main/java/com/refactoring/codesmells/dispensables/DuplicatedCode.java
package com.refactoring.codesmells.dispensables;

/**
 * Example of Duplicated Code smell.
 */
public class DuplicatedCode {

    public static class EmailService {

        public void sendWelcomeEmail(String email, String name) {
            // Duplicated validation
            if (email == null || email.isEmpty()) {
                throw new IllegalArgumentException("email is required");
            }
            if (!email.contains("@")) {
                throw new IllegalArgumentException("invalid email format");
            }

            // Duplicated formatting
            String subject = "Welcome!";
            String body = String.format("Hello %s, welcome to our service!", name);

            // Duplicated sending logic
            System.out.printf("Sending email to: %s%n", email);
            System.out.printf("Subject: %s%n", subject);
            System.out.printf("Body: %s%n", body);
            System.out.println("Email sent successfully");
        }

        public void sendPasswordResetEmail(String email, String resetToken) {
            // Duplicated validation
            if (email == null || email.isEmpty()) {
                throw new IllegalArgumentException("email is required");
            }
            if (!email.contains("@")) {
                throw new IllegalArgumentException("invalid email format");
            }

            // Duplicated formatting
            String subject = "Password Reset";
            String body = String.format("Use this token to reset your password: %s", resetToken);

            // Duplicated sending logic
            System.out.printf("Sending email to: %s%n", email);
            System.out.printf("Subject: %s%n", subject);
            System.out.printf("Body: %s%n", body);
            System.out.println("Email sent successfully");
        }

        public void sendOrderConfirmationEmail(String email, String orderId, double amount) {
            // Duplicated validation
            if (email == null || email.isEmpty()) {
                throw new IllegalArgumentException("email is required");
            }
            if (!email.contains("@")) {
                throw new IllegalArgumentException("invalid email format");
            }

            // Duplicated formatting
            String subject = "Order Confirmation";
            String body = String.format("Your order %s for $%.2f has been confirmed", orderId, amount);

            // Duplicated sending logic
            System.out.printf("Sending email to: %s%n", email);
            System.out.printf("Subject: %s%n", subject);
            System.out.printf("Body: %s%n", body);
            System.out.println("Email sent successfully");
        }
    }

    /*
     * Exercise: Refactor this code to eliminate duplication
     * 1. Extract the common validation logic
     * 2. Extract the common email sending logic
     * 3. Create a generic sendEmail method that the others can use
     */
}
